### DISCUSSION TITLE
Access local modules in a webpack library

### DISCUSSION THREAD

**Discussion by doberkofler** [webpack/webpack#13599] [Q&A]
I’m using a 3rd party library bundled using webpack. I would now need to use one of dependencies used by the library and would like to reuse the already loaded one instead of loading a new one. Is there a way to use one of the (not exported) local dependencies in a library bundled using webpack?

**Comment by alexander-akait** [comment]
Maybe you need this https://webpack.js.org/concepts/module-federation/

**Reply by doberkofler** [reply]
I had a look at this but I'm not sure if this really solves my use case. Could you please elaborate on the needed configuration in my webpack build to access the the local dependency in the 3rd party library?

**Reply by alexander-akait** [reply]
I recommend not to do this. Why you need this? Please describe use case

**Reply by doberkofler** [reply]
I using a 3rd party library that internally uses domvm fore the rendering. I must now also use domvm to inject markup in the virtual dom and would like to use the same version as used in the 3rd party library to make sure that I'm 100% compatible and no not boudle the same code twice.

**Reply by alexander-akait** [reply]
> 3rd party library

this library bundled by webpack?

**Reply by doberkofler** [reply]
yes

**Reply by alexander-akait** [reply]
Just reexport `domvm` in your library and use this from your library, so you can always have the same version of library and will be  100% compatible

**Reply by doberkofler** [reply]
I'm sorry, but I don't understand. The 3rd party library has domvm as a local dependency in the bundle. When I import the exported API from this 3rd party library I don't see the domvm exports and would therefore need install a new copy locally to import it in my local webpack bundle. Doing so would duplicate the domvm and I don't see how I cam guarantee that I'm using the exact same version.

**Reply by alexander-akait** [reply]
> Doing so would duplicate the domvm and I don't see how I cam guarantee that I'm using the exact same version.

So you don't use `domvm` directly in your library? In this case better setup CI lint rule or test and do it before bundle and publish your library. Modules inside bundled code will be shortened in production and have names like `a1b2c3`. Also checking compatibility is not really the task of webpack.

**Reply by alexander-akait** [reply]
Also if library uses `domvm`, it will be in `peerDependencies`, so developer can choose a version of `domvm`, for example look at `react` components https://github.com/tannerlinsley/react-table/blob/master/package.json#L55 (you can find them more using google). 

If developer don't specify `domvm` in `peerDependency` maybe he deliberately did it to be independent and don't forget that webpack do tree shaking and can remove some functions from `domvm`, so it is not safe to use directly

**Reply by doberkofler** [reply]
Yes, I do not use `domvm` in my own code.
Yes, the library vendor deliberately hides the dependency.

If I understand you correctly:
- it is generally speaking not possible to use a local dependency
- you would not recommend doing this even if possible
- you would load the `domvm` library again and manually deal with the version compatibility

**Reply by alexander-akait** [reply]
> it is generally speaking not possible to use a local dependency

in theory it is possible, you need run `__webpack_require__` directly with necessary module name, but don't forget in production module will have hashed name.

> you would not recommend doing this even if possible

Yes

> you would load the domvm library again and manually deal with the version compatibility

Yes

It makes sense to open an issue in library and ask why it is not in peer dependencies.

**Reply by doberkofler** [reply]
Out of interest: how would I use `__webpack_require__` ?

**Reply by alexander-akait** [reply]
Try to write `console.log(__webpack_modules__)` in your code and you will see modules
